### DISCUSSION TITLE
systemd, Startup Fails after Boot

### DISCUSSION THREAD

**Discussion by arrmo** [coredns/coredns#6288] [General]
Hi,

I have coredns set up to start on systemd, as a service. The service file is,
```
[Unit]
Description=CoreDNS DNS server
Documentation=https://coredns.io
After=network.target

[Service]
PermissionsStartOnly=true
LimitNOFILE=1048576
LimitNPROC=512
CapabilityBoundingSet=CAP_NET_BIND_SERVICE
AmbientCapabilities=CAP_NET_BIND_SERVICE
NoNewPrivileges=true
User=coredns
WorkingDirectory=~
ExecStart=/opt/coredns/coredns -conf=/etc/coredns/Corefile
ExecReload=/bin/kill -SIGUSR1 $MAINPID
Restart=on-failure

[Install]
WantedBy=multi-user.target
```

But ... after reboot, it always fails - does not start up. Manually restart it, no issues. Thinking perhaps it's a dependency of sorts? Any thoughts?

Thanks!

**Comment by polarathene** [comment]
> LimitNOFILE=1048576

Just random advice that you probably don't need that.

Since systemd v240 (2018Q2) the default should be `1024:524288`. The soft limit is `1024` for legacy/compatibility reasons, it shouldn't matter for Go I think (_at least pure Go programs_), and since Aug 2022 with Go 1.19+ those programs will implicitly raise soft limit to the hard limit, restoring the soft limit for any child processes (_bugfix earlier this year backported to Go 1.19 and 1.20_).

**Comment by arrmo** [comment]
Thanks! Tried it, no delta - thinking it may be that coredns needs to wait until the network connection is up? It says to wait on `network.target`, but is that sufficient?

**Reply by polarathene** [reply]
> but is that sufficient?

Not sure sorry, I don't write systemd service configs often.

You'd want to try look through logs to try identify what could be causing the issue. You could try introducing a delay like 5-10 sec and if that consistently starts then it's probably a race condition and you need that extra dependency identified.

**Comment by arrmo** [comment]
Will give it a try, thanks!

**Comment by Underknowledge** [comment]
lgtm, not sure if working directory ~ is valid thou 

that's what I use 
```bash
cat <<'EOF' > /etc/systemd/system/coredns.service
[Unit]
After=network.target
Description=coredns server
Documentation=http://coredns.io

[Service]
User=coredns
Type=simple
EnvironmentFile=/etc/sysconfig/coredns
ExecStart=/usr/bin/coredns --conf=/etc/coredns/Corefile $EXTRA_ARGS
ExecReload=/bin/kill -HUP $MAINPID
Restart=on-failure
RestartSec=10
SuccessExitStatus=0

[Install]
WantedBy=multi-user.target
EOF
echo 'EXTRA_ARGS=' > /etc/sysconfig/coredns 
mkdir -p /etc/systemd/system/coredns.service.d/
cat <<'EOF' > /etc/systemd/system/coredns.service.d/override.conf
[Service]
AmbientCapabilities=CAP_NET_BIND_SERVICE
CapabilityBoundingSet=CAP_NET_BIND_SERVICE
NoNewPrivileges=true
EOF
``` 

could it be selinux? 
`restorecon -v /usr/bin/coredns`
